### PR TITLE
Memory based loadbalancing

### PR DESCRIPTION
**Pull request by cbickel**
This PR implements the idea, that has been discussed in the following mail thread:
https://lists.apache.org/thread.html/dfccf972bc1419fe48dbc23119441108c45f85d53625fd6f8fc04fcb@%3Cdev.openwhisk.apache.org%3E

It changes the amount of containers an invoker can spawn based on available memory and not on the CPU.
In addition it makes the loadbalancer aware about the amount of available memory on each invoker and it limits the invoker to create only user containers, if there is enough free memory.

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [x] Loadbalancer
- [x] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by dgrove-oss** [comment]
Would love to see this get merged.  There are some compensating changes we will need in kube-deploy; I will be happy to take care of them when it is time.

File: tests/src/test/scala/whisk/core/containerpool/test/ContainerPoolTests.scala [hunk 1]
```diff
@@ -415,18 +573,18 @@ class ContainerPoolObjectTests extends FlatSpec with Matchers with MockFactory {
...
 
   it should "not provide a container if pool is empty" in {
-    ContainerPool.remove(Map()) shouldBe None
+    ContainerPool.remove(Map(), MemoryLimit.stdMemory) shouldBe List.empty
   }
 
   it should "not provide a container from busy pool with non-warm containers" in {
     val pool = Map('none -> noData(), 'pre -> preWarmedData())
-    ContainerPool.remove(pool) shouldBe None
+    ContainerPool.remove(pool, MemoryLimit.stdMemory) shouldBe List.empty
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/containerpool/test/ContainerPoolTests.scala]: Would it make sense to add a test which tests that if you cannot make space for enough capacity, the List is empty?

File: tests/src/test/scala/whisk/core/containerpool/test/ContainerPoolTests.scala [hunk 2]
```diff
@@ -124,7 +129,7 @@ class ContainerPoolTests
   it should "reuse a warm container" in within(timeout) {
     val (containers, factory) = testContainers(2)
     val feed = TestProbe()
-    val pool = system.actorOf(ContainerPool.props(factory, ContainerPoolConfig(2, 2), feed.ref))
+    val pool = system.actorOf(ContainerPool.props(factory, ContainerPoolConfig(MemoryLimit.stdMemory * 4), feed.ref))
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/containerpool/test/ContainerPoolTests.scala]: Please add a comment somewhere, that an action is created with stdMemory by default, so this is preserving behavior (4 actions can be scheduled)

File: tests/src/test/scala/whisk/core/containerpool/test/ContainerPoolTests.scala [hunk 3]
```diff
@@ -177,12 +182,35 @@ class ContainerPoolTests
     containers(1).expectMsg(runMessageDifferentEverything)
   }
 
+  it should "remove several containers to make space in the pool if it is already full and a different large action arrives" in within(
+    timeout) {
+    val (containers, factory) = testContainers(3)
+    val feed = TestProbe()
+
+    // a pool with slots for 512MB
+    val pool = system.actorOf(ContainerPool.props(factory, ContainerPoolConfig(512.MB), feed.ref))
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/containerpool/test/ContainerPoolTests.scala]: Should be `2 * stdMemory`

File: tests/src/test/scala/whisk/core/containerpool/test/ContainerPoolTests.scala [hunk 4]
```diff
@@ -177,12 +182,35 @@ class ContainerPoolTests
...
+    pool ! runMessage
+    containers(0).expectMsg(runMessage)
+    pool ! runMessageDifferentAction
+    containers(1).expectMsg(runMessageDifferentAction)
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/containerpool/test/ContainerPoolTests.scala]: Comments in here would be nice, like

```scala
containers(0).expectMsg(runMessage) // 1 * stdMemory taken
pool ! runMessageDifferentAction
containers(1).expectMsg(runMessageDifferentAction) // 2 * stdMemory taken -> full

...

pool ! runMessageLarge
// need to remove both action to make space for the large action (needs 2 * stdMemory)
containers(0).expectMsg(Remove)
containers(1).expectMsg(Remove)
containers(2).expectMsg(runMessageLarge)
```

File: tests/src/test/scala/whisk/core/containerpool/test/ContainerPoolTests.scala [hunk 5]
```diff
@@ -315,6 +383,96 @@ class ContainerPoolTests
...
+    feed.expectMsg(MessageFeed.Processed)
+
+    // Action 1 should start immediately
+    containers(0).expectMsgPF() {
+      // The `Some` assures, that it has been retried while the first action was still blocking the invoker.
+      case Run(runMessage.action, runMessage.msg, Some(_)) => true
+    }
+    // Action 2 should start immediately as well (without any retries, as there is already enough space in the pool)
+    containers(1).expectMsg(runMessageDifferentAction)
+  }
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/containerpool/test/ContainerPoolTests.scala]: Very nice test! 🎉

**Comment by cbickel** [comment]
PG5#488 🔵

**Comment by rabbah** [comment]
👀

File: core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala [hunk 6]
```diff
@@ -91,62 +98,87 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
...
+                // Only free up the amount, that is really needed to free up
+                  .remove(freePool, Math.min(r.action.limits.memory.megabytes, memoryConsumptionOf(freePool)).MB)
+                  .map(removeContainer)
+                  // If the list had at least one entry, enough containers were removed to start the new container. After
+                  // removing the containers, we are not interested anymore in the containers that have been removed.
+                  .headOption
+                  .map(_ =>
+                    takePrewarmContainer(r.action)
+                      .map(container => (container, "recreated prewarm"))
+                      .getOrElse(createContainer(r.action.limits.memory.megabytes.MB), "recreated")))
```
Comment (chetanmeh) [inline core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala]: Better to use non space separator like `recreated_prewarm` as the state gets used for Metric names. Currently following metric names get generated

* `openwhisk.counter.invoker_containerStart.warm_count`
* `openwhisk.counter.invoker_containerStart.cold_count`
* `openwhisk.counter.invoker_containerStart.precreated_count`

**Comment by cbickel** [comment]
The gatling-tests
`ApiV1Simulation`,
`LatencySimulation`,
`BlockingInvokeOneActionSimulation` and
`ColdBlockingInvokeSimulation`
do not show any performance regression on this PR.

File: common/scala/src/main/scala/whisk/core/entity/Size.scala [hunk 7]
```diff
@@ -69,6 +69,23 @@ case class ByteSize(size: Long, unit: SizeUnits.Unit) extends Ordered[ByteSize]
     ByteSize(commonSize, commonUnit)
   }
 
+  def *(other: Int): ByteSize = {
+    ByteSize(toBytes * other, SizeUnits.BYTE)
+  }
+
+  def /(other: ByteSize): Double = {
+    // Without throwing the exception the result would be `Infinity` here
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/Size.scala]: we could consider making this return a Try instead.

File: core/controller/src/main/scala/whisk/core/loadBalancer/ShardingContainerPoolBalancer.scala [hunk 8]
```diff
@@ -456,13 +461,15 @@ object ShardingContainerPoolBalancer extends LoadBalancerProvider {
    *
    * @param invokers a list of available invokers to search in, including their state
    * @param dispatched semaphores for each invoker to give the slots away from
+   * @param slots Number of slots, that need to be aquired (e.g. memory in MB)
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/loadBalancer/ShardingContainerPoolBalancer.scala]: acquired (typo)

File: common/scala/src/main/scala/whisk/core/entity/Size.scala (hunk 7 above)
Comment (cbickel) [reply common/scala/src/main/scala/whisk/core/entity/Size.scala]: What are the reasons for using a Try here?
On dividing `Int`s, you also get the response directly, instead of a `Try`, don't you?

**Comment by cbickel** [comment]
PG2#3520 🔵

File: core/controller/src/main/resources/reference.conf [hunk 9]
```diff
@@ -6,7 +6,7 @@ whisk {
     use-cluster-bootstrap: false
   }
   loadbalancer {
-    invoker-busy-threshold: 4
+    user-memory: 1024 m
```
Comment (chetanmeh) [inline core/controller/src/main/resources/reference.conf]: @cbickel Should this be named `invoker-user-memory`. I see following exception on startup as in my case `CONFIG_whisk_loadbalancer_invokerUserMemory` was not defined

```
Exception in thread "main" pureconfig.error.ConfigReaderException: Cannot convert configuration to a whisk.core.loadBalancer.ShardingContainerPoolBalancerConfig. Failures are:
  at 'whisk.loadbalancer':
    - Key not found: 'invoker-user-memory'.

	at pureconfig.package$.getResultOrThrow(package.scala:138)
	at pureconfig.package$.loadConfigOrThrow(package.scala:160)
	at whisk.core.loadBalancer.ShardingContainerPoolBalancer.<init>(ShardingContainerPoolBalancer.scala:159)
	at whisk.core.loadBalancer.ShardingContainerPoolBalancer$.instance(ShardingContainerPoolBalancer.scala:437)
	at whisk.core.controller.Controller.<init>(Controller.scala:117)
	at whisk.core.controller.Controller$.main(Controller.scala:258)
	at whisk.core.controller.Controller.main(Controller.scala)
```

Comment (cbickel) [reply core/controller/src/main/resources/reference.conf]: You are right.
I'll open a PR to correct this.
Thank you.

Comment (cbickel) [reply core/controller/src/main/resources/reference.conf]: #3993 has the fix.
